designs/limits.md

```diff
@@ -0,0 +1,60 @@
+
+To address the runaway-scaling problem the current fix in place is to detect if the kubelet for a worker node has never reported its status to the K8s control plane. If it's been longer than 15 minutes, Karpenter assumes that there's a hard failure mode due to which this worker node will never become healthy and terminates the worker node. If the condition map of the node object in the API Server says `NodeStatusNeverUpdated` then we use that as an indicator of the node having never come up.
+
+This fix ensures that if there are other scenarios where a worker node has become unhealthy due to a network partition or power outage in a availability zone, we don't terminate those worker nodes. It's important we don't make the static stability of a cluster worse during such an event. On the other hand, if there is an edge case where worker nodes come online and soon go offline, it will lead to runaway scaling again. This edge case should be unlikely to happen in the near term, so this document focuses on just the ability to limit costs within Karpenter. That way even if runaway scaling does occur there's a way to bound it. A longer-term solution to handle the runaway problem will be discussed separately.
+
+
+## Proposed Solution for Limits
+
+There are two broad forms of limiting we could apply. The first is that we could introduce a limit to the number of in-flight worker node being provisioned at a point in time. A worker node that's in the `NotReady` state could be considered to be in-flight. The second form is an absolute limit of the number of resources Karpenter can provision.
+
+
+### **In-flight limit**
+
+```yaml
+spec:
+  limits:
+    unready: 20% # StringOrInt
+```
+
+The good bit about this approach is that we don't artificially constrain how many total worker nodes can be spun up by Karpenter, while also making sure that if we're launching worker nodes that aren't healthy we stop the scaling and save costs.
+
+The two main problems with this approach though are -
+1. This limit while meant to constrain the number of unhealthy worker nodes in a cluster, will also inihibit the rate at which Karpenter can respond to pods that aren't schedulable. This somewhat goes against the goal of minimizing launch times of workers.
+2. While this helps ensure that costs don't increase due to runaway scaling, it won't help those who want a stricter cap on the amount of resources that's being provisioned even when nodes are otherwise healthy.
+
+### **Absolute limit**
+
+This is a much simpler semantic to follow, where you provide an upper cap on the amount of resources being provisioned irrespective of their health. This is more in-line with alternative solutions like CA and should be easier to reason with.
+
+The actual limit of the number of resources being spun up can be defined either through something similar to resource requests or just a flat count of instances.
+
+```yaml
+spec:
+  limits:
+    resources:
+      cpu: 1000
+      memory: 1000Gi
```

Comment:
I'm guessing you already had this in mind, but could you add GPU here as well?

Reply:
Yes will do